[Core] Add SGLang backend support in multi-process mode

DongDongJu opened the pull request:
**What this PR does / why we need it**:
This PR adds support for SGLang's KV cache format in LMCache, enabling SGLang to use LMCache as mp mode.

### Background
Related sglang PR : https://github.com/sgl-project/sglang/pull/16185
SGLang uses a different KV cache memory layout compared to vLLM:
- **vLLM MHA**: `[2, num_blocks, block_size, num_heads, head_size]` (combined K/V per layer)
- **vLLM MLA**: `[num_blocks, block_size, hidden_dim]` (combined K/V per layer)
- **SGLang MHA**: `[buffer_size, num_heads, head_size]` (separate K/V tensors, 2×num_layers total)

Due to this difference, SGLang requires the `multi_layer_kv_transfer_unilateral` kernel instead of `multi_layer_kv_transfer` for correct data transfer. 

### Changes

**1. Protocol Changes (`protocol.py`)**
- Add `backend_type: Optional[str]` parameter to `REGISTER_KV_CACHE` protocol
- Allows clients to explicitly specify their backend type ("sglang" or "vllm")
- Backward compatible: existing vLLM clients can omit the parameter (defaults to `None`)

**2. Server Changes (`server.py`)**

*GPUCacheContext:*
- Add `backend_type` parameter to constructor
- Add `is_sglang` property to identify SGLang format
- Handle SGLang's separate K/V tensor layout (2×num_layers tensors ordered as `[K0, K1, ..., V0, V1, ...]`)

*MPCacheEngine:*
- Update `register_kv_cache()` to accept `backend_type` parameter
- Modify `store()` to use `multi_layer_kv_transfer_unilateral` for SGLang
- Modify `retrieve()` to use `multi_layer_kv_transfer_unilateral` for SGLang

**Special notes for your reviewers**:

## Special notes for your reviewers

1. **Backward Compatibility**: All changes are backward compatible. Existing vLLM clients that don't pass `backend_type` will continue to work as before (defaults to vLLM format).

2. **Kernel Selection**: The key difference is which CUDA kernel is used:
   - vLLM: `multi_layer_kv_transfer` (for combined K/V buffers)
   - SGLang: `multi_layer_kv_transfer_unilateral` (for separate K/V buffers)

3. **Testing**: This has been tested with SGLang using `Qwen/Qwen3-0.6B` model with:
   - 6144 tokens stored and retrieved correctly
   - Deterministic output verification (temperature=0, top_k=1)

4. **SGLang Integration**: The corresponding SGLang PR implements `LMCacheMPConnector` that:
   - Passes `backend_type="sglang"` during registration
   - Orders KV tensors as `[K0, K1, ..., V0, V1, ...]` to match the unilateral kernel expectation

I tested with below three models and works well.
```
Qwen/Qwen3-0.6B
meta-llama/Meta-Llama-3.1-8B-Instruct
Qwen3-Coder-480B-A35B-Instruct-FP8 (pending)
```
## Test
using my sglang git (https://github.com/DongDongJu/sglang/tree/feature/lmcache-mp-support)
```bash
# 1. Start LMCache server
python3 -m lmcache.v1.multiprocess.server --port 5555 --cpu-buffer-size [mem size] --block-size 8

# 2. Start SGLang with MP mode
python3 -m sglang.launch_server \
    --model-path Qwen/Qwen3-0.6B \
    --port 30000 \
    --trust-remote-code \
    --tp [optional] \
    --ep [optional] \
    --lmcache-mp-enable \
    --random-seed 42 \
    --context-length 8192 \
    --lmcache-mp-host localhost \
    --lmcache-mp-port 5555

# 3. Send request to trigger STORE
curl -X POST http://localhost:30000/v1/completions \
    -H "Content-Type: application/json" \
    -d '{"model": "Qwen/Qwen3-0.6B", 'temperature': 0, 'top_k': 1, "prompt": "<long prompt>", "max_tokens": 50}'

# 4. Restart SGLang and send same request to trigger RETRIEVE
# Verify "Retrieved X tokens" in LMCache server logs
```

Commit:
[Protocol] Add backend_type parameter to REGISTER_KV_CACHE

Add an optional backend_type parameter to the REGISTER_KV_CACHE protocol
to allow clients to explicitly specify their backend type ("sglang" or
"vllm"). This enables the server to use the appropriate KV cache kernel
without relying on heuristics.

Changes:
- Add backend_type: Optional[str] to REGISTER_KV_CACHE payload
- Update protocol documentation

This is backward compatible - existing vLLM clients can omit the
parameter (defaults to None, which uses vLLM format).

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

Commit:
[Server] Add SGLang backend support in multi-process mode

Add support for SGLang's KV cache format in the multi-process server.
SGLang uses separate K/V tensors with shape [buffer_size, num_heads,
head_size], which requires the unilateral kernel for data transfer.

Changes in GPUCacheContext:
- Add backend_type parameter to explicitly identify the backend
- Add is_sglang property to check if using SGLang format
- Handle SGLang's separate K/V tensor layout (2*num_layers tensors)

Changes in MPCacheEngine:
- Update register_kv_cache to accept backend_type parameter
- Modify store() to use multi_layer_kv_transfer_unilateral for SGLang
- Modify retrieve() to use multi_layer_kv_transfer_unilateral for SGLang

The unilateral kernel is required for SGLang because:
- SGLang KV cache: [buffer_size, num_heads, head_size] (separate K/V)
- vLLM KV cache: [2, num_blocks, block_size, num_heads, head_size] (combined)

This change maintains full backward compatibility with vLLM clients.

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

Commit:
style fix

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>

Commit:
[Core] Add SGLang backend support in multi-process mode

This commit adds support for SGLang's KV cache format in LMCache's
multi-process mode by:

1. Protocol changes (protocol.py):
   - Add backend_type and block_size parameters to REGISTER_KV_CACHE
   - backend_type: "sglang" or "vllm" (None defaults to vLLM)
   - block_size: Required for SGLang backend slot mapping

2. Server changes (server.py):
   - GPUCacheContext now accepts explicit backend_type parameter
   - Use multi_layer_kv_transfer_unilateral kernel for SGLang format
   - Require block_size for SGLang backend (raise error if not provided)
   - Reorder KV cache pointers for SGLang's separate K/V tensors

3. Test changes (test_cache_server.py):
   - Update tests to pass backend_type and block_size parameters

SGLang uses separate K and V tensors with shape [buffer_size, num_heads,
head_size], requiring the unilateral kernel instead of the combined
K/V kernel used by vLLM.

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu commented on lmcache/v1/multiprocess/server.py:
```diff
@@ -54,29 +55,79 @@ def list_to_gpu_tensor(lis: list[int], device: torch.device) -> torch.Tensor:
+            # The block_size is passed when registering, but we don't have it here
+            # Use a reasonable default (8 is common)
+            self.block_size_ = 8  # Will be updated if needed
+            self.num_blocks_ = buffer_size // self.block_size_
```
I addded block_size param in https://github.com/LMCache/LMCache/pull/2333/commits/6e4ac85889b2ceb5e1e1372cc9d965cfa24963a6

Commit:
[Tests] Update test handlers for new REGISTER_KV_CACHE signature

Update test_mq.py and test_mq_handler_helpers.py to pass the new
backend_type and block_size parameters added to REGISTER_KV_CACHE.

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

ApostaC commented:
Hey @DongDongJu , it would be better if we could converge on the design first. 
I would say we should decouple the serving-engine-related processing from the multi-process server itself, and let's try to keep the IPC interface as stable as possible, and make it agnostic to the serving engines.

DongDongJu commented:
> Hey @DongDongJu , it would be better if we could converge on the design first. I would say we should decouple the serving-engine-related processing from the multi-process server itself, and let's try to keep the IPC interface as stable as possible, and make it agnostic to the serving engines.

Make sense. I will bring design notes for the discussion in next community meeting.

DongDongJu committed:
Merge branch 'dev' into feature/sglang-mp-support

DongDongJu committed:
Merge branch 'LMCache:dev' into feature/sglang-mp-support

Commit:
feat: Add KV cache format support for engine-agnostic registration

- Add KVCacheFormat and related types for describing KV cache layouts
- Support MHA and MLA format detection
- Enable format-based registration for SGLang and vLLM backends

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

Commit:
feat: Extend custom types for KV cache format support

- Add KVCacheFormat to custom types for serialization
- Update type definitions to support format-based registration

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

Commit:
feat: Extend protocol to support KV cache format in registration

- Add optional KVCacheFormat parameter to REGISTER_KV_CACHE
- Support backward compatibility with legacy registration

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

Commit:
feat: Update message queue for format-based registration

- Handle KVCacheFormat in message serialization
- Support format parameter in registration requests

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

Commit:
feat: Add SGLang support to LMCache multiprocess server

- Support SGLang backend type detection and format-based registration
- Add KV cache format handling for MHA and MLA layouts
- Implement format-agnostic GPU context initialization
- Add hash configuration verification for consistent caching
- Support both legacy and format-based registration modes

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>
Signed-off-by: DongDongJu <commisori28@gmail.com>

DongDongJu committed:
Merge branch 'dev' into feature/sglang-mp-support

Commit:
fix: re-export KVLayerGroupSpec and L0LayoutSpec in custom_types

Add KVLayerGroupSpec and L0LayoutSpec to __all__ to prevent ruff from
removing them as unused imports. These types are imported by
sglang's multi_process_adapter module and need to be available
from lmcache.v1.multiprocess.custom_types.

Signed-off-by: Dongjoo Seo <dongjoo.seo1@samsung.com>

DongDongJu commented:
will reopen after mp mode redesign is done. maybe next week?
